File: typescript/src/main/vsdk/toolkit/galerkin/processing/GatheringStrategy.ts
import { RenderOptions } from "../../common/RenderOptions";
import { GalerkinElement } from "../GalerkinElement";
import { GalerkinState } from "../GalerkinState";
import { Scene } from "../../scene/Scene";

export abstract class GatheringStrategy {
  protected static pushPullPotential(element: GalerkinElement | null, down: number): number {
    if (element === null) {
      return 0.0;
    }

    down += element.area > 0.0 ? (element.receivedPotential / element.area) : 0.0;
    element.receivedPotential = 0.0;
    let up = 0.0;

    if (element.regularSubElements === null && element.irregularSubElements === null) {
      up = down + (element.patch !== null ? element.patch.directPotential : 0.0);
    }

    if (element.regularSubElements !== null) {
      for (let i = 0; i < 4; i++) {
        if (element.regularSubElements[i] instanceof GalerkinElement) {
          up += 0.25 * GatheringStrategy.pushPullPotential(element.regularSubElements[i] as GalerkinElement, down);
        }
      }
    }

    if (element.irregularSubElements !== null) {
      for (let j = 0; j < element.irregularSubElements.length; j++) {
        if (!(element.irregularSubElements[j] instanceof GalerkinElement)) {
          continue;
        }
        const subElement = element.irregularSubElements[j] as GalerkinElement;
        const localDown = element.isCluster() ? down : 0.0;
        up += element.area > 0.0
          ? (subElement.area / element.area) * GatheringStrategy.pushPullPotential(subElement, localDown)
          : 0.0;
      }
    }

    element.potential = up;
    return up;
  }

  public constructor() {
  }

  public abstract doGatheringIteration(scene: Scene, galerkinState: GalerkinState, renderOptions: RenderOptions): boolean;
}

File: java/src/main/java/vsdk/toolkit/galerkin/processing/GatheringStrategy.java
package vsdk.toolkit.galerkin.processing;

import vsdk.toolkit.material.RenderOptions;
import vsdk.toolkit.galerkin.GalerkinElement;
import vsdk.toolkit.galerkin.GalerkinState;
import vsdk.toolkit.scene.Scene;

/**
Reference:
[COHE1993] Cohen, M. Wallace, J. "Radiosity and Realistic Image Synthesis",
     Academic Press Professional, 1993.
*/

/**
Numerical integration for Jacobi or Gauss-Seidel Galerkin radiosity.
See [COHE1993] section 5.3.2.
*/
public abstract class GatheringStrategy {
    protected static float pushPullPotential(GalerkinElement element, float down) {
        if ( element == null ) {
            return 0.0f;
        }
        down += element.area > 0.0f ? (element.receivedPotential / element.area) : 0.0f;
        element.receivedPotential = 0.0f;
        float up = 0.0f;

        if ( element.regularSubElements == null && element.irregularSubElements == null ) {
            up = down + (element.patch != null ? element.patch.directPotential : 0.0f);
        }

        if ( element.regularSubElements != null ) {
            for ( int i = 0; i < 4; i++ ) {
                if ( element.regularSubElements[i] instanceof GalerkinElement ) {
                    up += 0.25f * pushPullPotential((GalerkinElement)element.regularSubElements[i], down);
                }
            }
        }

        if ( element.irregularSubElements != null ) {
            for ( int j = 0; j < element.irregularSubElements.size(); j++ ) {
                if ( !(element.irregularSubElements.get(j) instanceof GalerkinElement) ) {
                    continue;
                }
                GalerkinElement subElement = (GalerkinElement)element.irregularSubElements.get(j);
                float localDown = element.isCluster() ? down : 0.0f;
                up += element.area > 0.0f
                    ? (subElement.area / element.area) * pushPullPotential(subElement, localDown)
                    : 0.0f;
            }
        }

        element.potential = up;
        return up;
    }

    public GatheringStrategy() {
    }

    public abstract boolean doGatheringIteration(Scene scene, GalerkinState galerkinState, RenderOptions renderOptions);
}
